docscan/session/retrieve/get_session_result.go

```diff
@@ -1,12 +1,72 @@
+	"fmt"
+
+	"github.com/getyoti/yoti-go-sdk/v3/docscan/constants"
+)
+
 // GetSessionResult contains the information about a created session
 type GetSessionResult struct {
-	ClientSessionTokenTTL int               `json:"client_session_token_ttl"`
-	ClientSessionToken    string            `json:"client_session_token"`
-	SessionID             string            `json:"session_id"`
-	UserTrackingID        string            `json:"user_tracking_id"`
-	State                 string            `json:"state"`
-	Checks                []CheckResponse   `json:"checks"`
-	Resources             ResourceContainer `json:"resources"`
+	ClientSessionTokenTTL int                `json:"client_session_token_ttl"`
+	ClientSessionToken    string             `json:"client_session_token"`
+	SessionID             string             `json:"session_id"`
+	UserTrackingID        string             `json:"user_tracking_id"`
+	State                 string             `json:"state"`
+	Checks                []*CheckResponse   `json:"checks"`
+	Resources             *ResourceContainer `json:"resources"`
+	authenticityChecks    []*AuthenticityCheckResponse
+	faceMatchChecks       []*FaceMatchCheckResponse
+	textDataChecks        []*TextDataCheckResponse
+	livenessChecks        []*LivenessCheckResponse
+}
+
+// AuthenticityChecks filters the checks, returning only document authenticity checks
+func (g *GetSessionResult) AuthenticityChecks() []*AuthenticityCheckResponse {
+	return g.authenticityChecks
+}
+
+// FaceMatchChecks filters the checks, returning only FaceMatch checks
+func (g *GetSessionResult) FaceMatchChecks() []*FaceMatchCheckResponse {
+	return g.faceMatchChecks
+}
+
+// TextDataChecks filters the checks, returning only Text Data checks
+func (g *GetSessionResult) TextDataChecks() []*TextDataCheckResponse {
+	return g.textDataChecks
+}
+
+// LivenessChecks filters the checks, returning only Liveness checks
+func (g *GetSessionResult) LivenessChecks() []*LivenessCheckResponse {
+	return g.livenessChecks
+}
+
+// UnmarshalJSON handles the custom JSON unmarshalling
+func (g *GetSessionResult) UnmarshalJSON(data []byte) error {
+	type result GetSessionResult // declared as "type" to prevent recursive unmarshalling
+	if err := json.Unmarshal(data, (*result)(g)); err != nil {
+		return err
+	}
+
+	for _, check := range g.Checks {
+		switch check.Type {
+		case constants.IDDocumentAuthenticity:
+			g.authenticityChecks = append(g.authenticityChecks, &AuthenticityCheckResponse{CheckResponse: check})
+
+		case constants.IDDocumentFaceMatch:
+			g.faceMatchChecks = append(g.faceMatchChecks, &FaceMatchCheckResponse{CheckResponse: check})
+
+		case constants.IDDocumentTextDataCheck:
+			g.textDataChecks = append(g.textDataChecks, &TextDataCheckResponse{CheckResponse: check})
+
+		case constants.Liveness:
+			g.livenessChecks = append(g.livenessChecks, &LivenessCheckResponse{CheckResponse: check})
+
+		default:
+			fmt.Printf("Unrecognized check type: `%s`", check.Type)
```

Comment:
It is my understanding that it's considered bad form for an SDK to write directly to std out. I'm not really sure what to do in this case, it might be safer to either return an error, or ignore the case. Since it's possible to get all the checks even if this fails, I'm leaning towards just ignoring.

Reply:
Other parts of the SDK use `log.Printf`, so perhaps `fmt.Printf` wasn't intended here?

If this is the case, does it make sense to use `log.Printf` here (and in other parts of SDK)? If not, perhaps we should review it use throughout the SDK as a separate task?

Reply:
Opinions may vary here, but I would not log in the SDK.

Integrators could be using std err and std out for their own logging and mandate certain formatting that would then be screwed up by the SDK coming along and do it's own thing.

I know I have struggled with similar things in other projects.

Reply:
One way around could be to support a custom logger interface which integrators can use if they want more detailed information, but it might be overkill in this case. You're probably better off doing a best effort in handling unexpected situations, and if you fail return an error.

Reply:
I'm unsure if we want to be returning an error in this instance though if there's an unrecognised check. If we add a new check in future, we don't want to stop clients from being able to process the other checks, which might happen if we throw an error.

On the other hand, they surely would have configured this check to come through, so maybe we should be throwing an error, since they should be expecting it.
Writing this, I think I'm leaning towards that we should be returning an error.

Reply:
Regarding the logging, I guess you could argue we should always be returning that information as an error when possible. So we can revisit that as a separate task

Reply:
In other SDKs, we just return a generic `CheckResponse`, e.g.
- https://github.com/getyoti/yoti-java-sdk/blob/master/yoti-sdk-impl/src/main/java/com/yoti/api/client/docs/session/retrieve/SimpleCheckResponse.java#L16
- https://github.com/getyoti/yoti-php-sdk/blob/master/src/DocScan/Session/Retrieve/GetSessionResult.php#L175

Reply:
I think we should skip without logging or returning an error. An new/unknown check will be available in `Checks` and will have a `Type`

Reply:
You could do a Generic Check Response, but you in a way already have that in that you expose the Checks slice which an SDK user can defer to if they have a check that is not supported in their version of the SDK.

Reply:
Yes they can get their check from the generic Checks slice, so we'll be fine. I've made an update in 1a2d3f9 to remove this logging. We'll look at the logging in other places in the SDK before we release the major